File: json/json-annotation-processor/src/test/java/ru/tinkoff/kora/json/annotation/processor/dto/DtoOnlyReader.java
package ru.tinkoff.kora.json.annotation.processor.dto;

import ru.tinkoff.kora.json.common.annotation.JsonField;
import ru.tinkoff.kora.json.common.annotation.JsonReader;
import tools.jackson.core.JsonParser;
import tools.jackson.core.JsonToken;

@JsonReader
public record DtoOnlyReader(
    String field1,
    @JsonField("renamedField2") String field2,
    @JsonField(reader = Field3Reader.class) Inner field3
) {

    public static final class Field3Reader implements ru.tinkoff.kora.json.common.JsonReader<Inner> {

        @Override
        public Inner read(JsonParser parser) {
            var token = parser.currentToken();
            if (token != JsonToken.VALUE_STRING) {
                throw new RuntimeException();
            }
            var value = parser.getValueAsString();
            return new Inner(value);
        }
    }

    public record Inner(String value) {
    }
}

File: json/json-symbol-processor/src/test/kotlin/ru/tinkoff/kora/json/ksp/dto/DtoOnlyReader.kt
package ru.tinkoff.kora.json.ksp.dto

import ru.tinkoff.kora.json.common.annotation.JsonField
import ru.tinkoff.kora.json.common.annotation.JsonReader
import tools.jackson.core.JsonParser
import tools.jackson.core.JsonToken

@JsonReader
data class DtoOnlyReader(
    val field1: String,
    @JsonField("renamedField2") val field2: String,
    @JsonField( reader = Field3Reader::class) val field3: Inner
) {
    class Field3Reader : ru.tinkoff.kora.json.common.JsonReader<Inner?> {
        override fun read(parser: JsonParser): Inner {
            val token = parser.currentToken()
            if (token != JsonToken.VALUE_STRING) {
                throw RuntimeException()
            }
            val value = parser.valueAsString
            return Inner(value)
        }
    }

    data class Inner(val value: String)
}
